pkg/cloud/services/network/secondarycidr_test.go

```diff
@@ -0,0 +1,264 @@
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name              string
+		haveSecondaryCIDR bool
+		expect            func(m *mock_ec2iface.MockEC2APIMockRecorder)
+		wantErr           bool
+	}{
+		{
+			name: "Should not associate secondary CIDR if no secondary cidr block info present in control plane",
+		},
+		{
+			name:              "Should return error if unable to describe VPC",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(nil, awserrors.NewFailedDependency("dependency-failure"))
+			},
+			wantErr: true,
+		},
+		{
+			name:              "Should return error if no VPC found",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(nil, nil)
+			},
+			wantErr: true,
+		},
+		{
+			name:              "Should not associate secondary cidr block if already exist in VPC",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(&ec2.DescribeVpcsOutput{
+					Vpcs: []*ec2.Vpc{
+						{
+							CidrBlockAssociationSet: []*ec2.VpcCidrBlockAssociation{
+								{CidrBlock: aws.String("secondary-cidr")},
+							},
+						},
+					}}, nil)
+			},
+		},
+		{
+			name:              "Should associate secondary cidr block if not exist in VPC",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(&ec2.DescribeVpcsOutput{
+					Vpcs: []*ec2.Vpc{
+						{
+							CidrBlockAssociationSet: []*ec2.VpcCidrBlockAssociation{
+								{CidrBlock: aws.String("secondary-cidr-new")},
+							},
+						},
+					}}, nil)
+				m.AssociateVpcCidrBlock(gomock.AssignableToTypeOf(&ec2.AssociateVpcCidrBlockInput{})).Return(nil, nil)
+			},
+		},
+		{
+			name:              "Should return error if failed during associating secondary cidr block",
+			haveSecondaryCIDR: true,
+			expect: func(m *mock_ec2iface.MockEC2APIMockRecorder) {
+				m.DescribeVpcs(gomock.AssignableToTypeOf(&ec2.DescribeVpcsInput{})).Return(&ec2.DescribeVpcsOutput{
+					Vpcs: []*ec2.Vpc{
+						{
+							CidrBlockAssociationSet: []*ec2.VpcCidrBlockAssociation{
+								{CidrBlock: aws.String("secondary-cidr-new")},
+							},
+						},
+					}}, nil)
+				m.AssociateVpcCidrBlock(gomock.AssignableToTypeOf(&ec2.AssociateVpcCidrBlockInput{})).Return(nil, awserrors.NewFailedDependency("dependency-failure"))
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			scheme, err := setupScheme()
+			g.Expect(err).NotTo(HaveOccurred())
+			cl := fake.NewClientBuilder().WithScheme(scheme).Build()
```

Comment:
Since this client is used only in setupNewManagedControlPlaneScope func, can we declare it inside the func?

Reply:
We could do this, but the responsibility of `setupNewManagedControlPlaneScope` is to just return the managedControlPlaneScope object. And I have kept the schemeSetup and client creation out of this scope.
And from my previous experience I have observed that in test cases we might need to create some other scopes also like machineScope etc. So in those scenarios we can reuse the same client. That's why not merged it in the same func.

Reply:
IMO if any variable is used only for a particular func w should keep it in the function scope, so if at all it's required in future that other scopes may come into picture, we can always refactor it back.

Reply:
I got your point regarding variable scope, but responsibility of setupNewManagedControlPlaneScope() is to create managedControlPlaneScope object. Adding the creation of client inside this function seems against single responsibility principle to me. Do you still think client should be created inside the method?